lib/TPP/Transforms/VectorContractToOuterproduct.cpp

```diff
@@ -0,0 +1,247 @@
+  auto results = map.getResults();
+  if (results.size() != matrixDims)
+    return false;
+
+  // Check the last two dimensions for transposition
+  auto secondLast = dyn_cast<AffineDimExpr>(results[matrixDims - 2]);
+  auto last = dyn_cast<AffineDimExpr>(results[matrixDims - 1]);
+
+  if (!secondLast || !last)
+    return false;
+
+  // For LHS, If the last dimension comes before the second last, it's
+  // transposed. its opposite for RHS.
+  return isRHS ? last.getPosition() > secondLast.getPosition()
+               : last.getPosition() < secondLast.getPosition();
+}
+
+struct VectorContractToOuterproductPattern
+    : public OpRewritePattern<vector::ContractionOp> {
+  using OpRewritePattern<vector::ContractionOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::ContractionOp contractOp,
+                                PatternRewriter &rewriter) const override {
+
+    if (contractOp.getKind() != vector::CombiningKind::ADD)
+      return rewriter.notifyMatchFailure(
+          contractOp,
+          "Unsupported combining kind, only supports ADD at the moment)");
+
+    auto maps = contractOp.getIndexingMapsArray();
+    if (llvm::any_of(
+            maps, [](AffineMap map) { return !map.isProjectedPermutation(); }))
+      return rewriter.notifyMatchFailure(contractOp, "Unexpected map");
+
+    // Check if this is a matrix multiply
+    auto iteratorTypes = contractOp.getIteratorTypesArray();
+    if (iteratorTypes.size() < 3 ||
+        iteratorTypes[iteratorTypes.size() - 3] !=
+            vector::IteratorType::parallel ||
+        iteratorTypes[iteratorTypes.size() - 2] !=
+            vector::IteratorType::parallel ||
+        iteratorTypes[iteratorTypes.size() - 1] !=
+            vector::IteratorType::reduction)
+      return rewriter.notifyMatchFailure(contractOp, "Not a gemm");
+
+    Location loc = contractOp.getLoc();
+    Value lhs = contractOp.getLhs();
+    Value rhs = contractOp.getRhs();
+    Value acc = contractOp.getAcc();
+
+    // Find the original tensor operands
+    auto lhsDefiningOp = lhs.getDefiningOp<vector::TransferReadOp>();
+    auto rhsDefiningOp = rhs.getDefiningOp<vector::TransferReadOp>();
+    auto accDefiningOp = acc.getDefiningOp<vector::TransferReadOp>();
+    if (!lhsDefiningOp || !rhsDefiningOp || !accDefiningOp)
+      return failure();
+
+    // Make sure the inputs being read are whole tensor or subview.
+    if (!llvm::all_of(lhsDefiningOp.getIndices(), isZeroIndex) ||
+        !llvm::all_of(rhsDefiningOp.getIndices(), isZeroIndex)) {
+      return failure();
+    }
+
+    Value lhsTensor = lhsDefiningOp.getSource();
+    Value rhsTensor = rhsDefiningOp.getSource();
+
+    auto lhsType = cast<ShapedType>(lhsDefiningOp.getType());
+    auto rhsType = cast<ShapedType>(rhsDefiningOp.getType());
+    auto accType = cast<ShapedType>(accDefiningOp.getType());
+
+    // Only 2-D output expected.
+    if (accType.getRank() != 2)
```

Comment:
Don't we expect all types to be 2D?

Reply:
No, as we may be working on higher dimension subviews as shown in the example above.

Reply:
But in the case of the `batch_matmul`, the output is 3D too, right? So this check (and the checks on the inputs) need to be conditional on what flavour of GEMM we're doing.

Reply:
Yes, updated.